server/src/main/java/org/elasticsearch/action/admin/indices/datastream/DataStreamsStatsAction.java

```diff
@@ -0,0 +1,377 @@
+            return shardRouting;
+        }
+
+        public StoreStats getStoreStats() {
+            return storeStats;
+        }
+
+        public long getMaxTimestamp() {
+            return maxTimestamp;
+        }
+    }
+
+    private static class AggregatedStats {
+        Set<String> backingIndices = new HashSet<>();
+        long storageBytes = 0L;
+        long maxTimestamp = 0L;
+    }
+
+    public static class TransportAction extends TransportBroadcastByNodeAction<Request, Response, DataStreamShardStats> {
+
+        private final ClusterService clusterService;
+        private final IndicesService indicesService;
+        private final IndexAbstractionResolver indexAbstractionResolver;
+
+        @Inject
+        public TransportAction(ClusterService clusterService, TransportService transportService, IndicesService indicesService,
+                                               ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+            super(DataStreamsStatsAction.NAME, clusterService, transportService, actionFilters, indexNameExpressionResolver,
+                Request::new, ThreadPool.Names.MANAGEMENT);
+            this.clusterService = clusterService;
+            this.indicesService = indicesService;
+            this.indexAbstractionResolver = new IndexAbstractionResolver(indexNameExpressionResolver);
+        }
+
+        @Override
+        protected Request readRequestFrom(StreamInput in) throws IOException {
+            return new Request(in);
+        }
+
+        @Override
+        protected ClusterBlockException checkGlobalBlock(ClusterState state, Request request) {
+            return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+        }
+
+        @Override
+        protected ClusterBlockException checkRequestBlock(ClusterState state, Request request, String[] concreteIndices) {
+            return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, concreteIndices);
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState clusterState, Request request, String[] concreteIndices) {
+            String[] requestIndices = request.indices();
+            if (requestIndices == null || requestIndices.length == 0) {
+                requestIndices = new String[]{"*"};
+            }
+            List<String> abstractionNames = indexAbstractionResolver.resolveIndexAbstractions(requestIndices, request.indicesOptions(),
+                clusterState.getMetadata());
+            SortedMap<String, IndexAbstraction> indicesLookup = clusterState.getMetadata().getIndicesLookup();
+
+            String[] concreteDatastreamIndices = abstractionNames.stream().flatMap(abstractionName -> {
+                IndexAbstraction indexAbstraction = indicesLookup.get(abstractionName);
+                assert indexAbstraction != null;
+                if (indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM) {
```

Comment:
Maybe we should fail the request if anything other than a data stream is specified?

Reply:
Perhaps this makes sense if we receive a concrete expression that is a non-wildcard input, but `*` is a valid input here which would technically match indices.

Reply:
> Perhaps this makes sense if we receive a concrete expression that is a non-wildcard input, but `*` is a valid input here which would technically match indices.

I have a similar issue on a PR that I have open for converting data stream admin actions from cluster-level actions to index-level actions (https://github.com/elastic/elasticsearch/pull/59095). Those actions should be applied to anything other than data streams but with security enabled, wildcards are expanded to indices, aliases and data streams. We solved the reverse problem (don't expand wildcards for index actions that cannot operate on data streams to include data streams) but it seems like in a few cases, we need the opposite -- don't expand wildcards on actions that support only data streams to include indices and aliases.

Reply:
>  if we receive a concrete expression that is a non-wildcard input, but * is a valid input here which would technically match indices.

I see, I forgot about that. In that case it makes sense to not fail and just filter out the `IndexAbstractions` that are not data streams.